packages/chat-ui/src/hook/use-chat-workflow/helper.ts

```diff
@@ -105,3 +177,51 @@ function toVercelAnnotations(event: WorkflowEvent) {
     }
   }
 }
+
+function convertRawNodeToSourceNode(
+  rawNode: RawNode,
+  fileServerUrl?: string
+): SourceNode {
+  const { node, score } = rawNode
+
+  return {
+    id: node.id_,
+    metadata: node.metadata,
+    score,
+    text: node.text,
+    url: getDocumentUrlFromRawNode(rawNode, fileServerUrl),
+  }
+}
+
+function getDocumentUrlFromRawNode(
+  rawNode: RawNode,
+  fileServerUrl?: string
+): string {
+  const { metadata } = rawNode.node
+  const { URL: fileURL, file_name, pipeline_id } = metadata
+
+  // if the file URL is provided in the metadata, use it
+  if (fileURL) return fileURL
+
+  // if `fileServerUrl` is not provided, return empty string
+  if (!fileServerUrl) {
+    console.warn(
+      `No 'fileServerUrl' provided. Nodes won't be displayed in ChatSources UI.`
+    )
+    return ''
+  }
+
+  // check if file_name is provided in the metadata
+  if (!file_name) {
+    console.warn(
+      `No file name found in this raw node: ${JSON.stringify(rawNode)}. It won't be displayed in ChatSources UI.`
+    )
+    return ''
+  }
+
+  if (pipeline_id) {
+    return `${fileServerUrl}/output/llamacloud/${pipeline_id}$${file_name}`
+  }
+
+  return `${fileServerUrl}/data/${file_name}`
```

Comment:
how about we let the user of the component choose the complete file server url?
```suggestion
  return `${fileServerUrl}/${file_name}`
```